aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h

```diff
@@ -0,0 +1,256 @@
+#pragma once
+
+#include <ATen/Tensor.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/Dispatch.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/arange.h>
+#include <ATen/ops/empty.h>
+#include <ATen/ops/_sparse_coo_tensor_with_dims_and_tensors.h>
+#endif
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+#define GPUCC
+#define FUNCAPI __host__ __device__
+#define INLINE __forceinline__
+#define NAME "compressed_index_invariance_checks_cuda"
+#else
+#define FUNCAPI
+#define INLINE inline
+#define NAME "compressed_index_invariance_checks_cpu"
+#endif
+
+#if defined(_WIN32) || defined(_WIN64)
+#define RESTRICT __restrict
+#else
+#define RESTRICT __restrict__
+#endif
+
+#define INVARIANT_CHECK_FUNC_API static INLINE FUNCAPI void
+
+namespace at {
+namespace native {
+
+namespace {
+
+// NOTE: all the checks but the very last one are designed
+// to work with vectors.
+// To enable vectorization one would need to write a conversion
+// Vec -> bool and make kernel launchers call into vectorized
+// execution paths.
+
+// All the invariants are described in https://pearu.github.io/bsr_tensor_invariants.html
+
+INVARIANT_CHECK_FUNC_API
+_assert(const bool cond, const char* const message) {
+#ifdef GPUCC
+  CUDA_KERNEL_ASSERT(cond && message);
+#else
+  TORCH_CHECK(cond, message);
+#endif
+}
+
+// Invariant 5.1
+// cidx[..., 0] == 0
```

Comment:
@cpuhrsch , with @pearu we decided to change these to `plain_indices` and `compressed_indices`, as a middle ground between performance and UX. What do you think about making these names available as methods for compressed formats? These could save as from code duplications provided that they are well-documented for the user. The user will benefit from them as well to write generic code that works both for CSR and CSR at the same time.

Reply:
@nikitaved - Once we have the documentation written up I think we can add them. However, in the context of the unified data model we likely want to remove these again in the future. cc @bhosmer